at_client/lib/src/service/sync_service_impl.dart

```diff
@@ -45,76 +53,121 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
-          onError(syncResult);
-        }
-        return syncResult;
+      if (await isInSync()) {
+        _logger.finer('server and local are in sync - ${syncRequest._id}');
+        syncRequest.result!
+          ..syncStatus = SyncStatus.success
+          ..lastSyncedOn = DateTime.now().toUtc();
+        _syncComplete(syncRequest);
+        return;
       }
-      var serverCommitId;
-      try {
-        // Check if local and cloud secondary are in sync. If true, return.
-        if (await isInSync()) {
-          syncResult = SyncResult();
-          _logger.info('Local Secondary and Cloud Secondary are in sync');
-          // Setting isSyncInProgress to false, to allow next sync call.
-          _isSyncInProgress = false;
-          if (onDone != null) {
-            onDone(syncResult);
-          }
-          return syncResult;
-        }
-        // Get latest server commit id.
-        serverCommitId = await _getServerCommitId();
-      } on AtLookUpException catch (exception) {
-        _logger.severe(
-            '${_atClient.getCurrentAtSign()} ${exception.errorMessage}');
-        syncResult = SyncResult();
-        syncResult.syncStatus = SyncStatus.failure;
-        syncResult.atClientException =
-            AtClientException(exception.errorCode, exception.errorMessage);
-        // Setting isSyncInProgress to false, to allow next sync call.
-        _isSyncInProgress = false;
-        if (onError != null) {
-          onError(syncResult);
-        }
-        return syncResult;
-      }
-      // Sync
-      return await _sync(serverCommitId, onDone, onError);
-    } finally {
-      _isSyncInProgress = false;
+    } on AtLookUpException catch (e) {
+      _logger.severe(
+          'Atlookup exception for sync ${syncRequest._id}. Reason ${e.toString()}');
+      syncRequest.result!.atClientException =
+          AtClientException(e.errorCode, e.errorMessage);
+      _syncError(syncRequest);
+      return;
+    }
+    final serverCommitId = await _getServerCommitId();
+    await _sync(serverCommitId, syncRequest);
+    _syncComplete(syncRequest);
+  }
+
+  void _syncError(SyncRequest syncRequest) {
+    if (syncRequest.onError != null) {
+      syncRequest.onError!(syncRequest.result);
+    }
+  }
+
+  void _syncComplete(SyncRequest syncRequest) {
+    syncRequest.result!.lastSyncedOn = DateTime.now().toUtc();
+    if (syncRequest.onDone != null) {
+      syncRequest.onDone!(syncRequest.result);
+    }
+    _clearQueue();
+  }
+
+  void _onDone(SyncResult syncResult) {
+    _logger.finer('system sync completed on ${syncResult.lastSyncedOn}');
+  }
+
+  void _onError(SyncResult syncResult) {
+    _logger.severe(
+        'system sync error ${syncResult.atClientException?.errorMessage}');
+  }
+
+  void _addSyncRequestToQueue(SyncRequest syncRequest) {
+    if(_syncRequests.length == _queueSize) {
+      _syncRequests.removeLast();
     }
+    _syncRequests.addLast(syncRequest);
+    syncRequest.result = SyncResult();
```

Comment:
ideally, this should happen elsewhere. This is a side effect. i.e doing beyond adding to queue.

Reply:
done